LeanOA/ForMathlib/Analysis/SpecialFunctions/ContinuousFunctionalCalculus/Abs.lean

```diff
@@ -0,0 +1,285 @@
+variable [NonUnitalContinuousFunctionalCalculus ℝ≥0 (fun (a : A) => 0 ≤ a)]
+
+/-- The absolute value of an operator, using the nonunital continuous functional calculus. -/
+noncomputable def abs (a : A) := sqrt (star a * a)
+
+@[simp]
+lemma abs_nonneg {a : A} : 0 ≤ abs a := sqrt_nonneg
+
+@[simp]
+lemma abs_zero : abs (0 : A) = 0 := by
+  rw [abs, star_zero, mul_zero, sqrt_zero]
+
+-- It's likely that using these variables warrants a different section
+
+variable [StarOrderedRing A] [UniqueNonUnitalContinuousFunctionalCalculus ℝ≥0 A]
+
+lemma abs_mul_self (a : A) : (abs a) * (abs a) = star a * a := by
+  refine sqrt_mul_sqrt_self _ <| star_mul_self_nonneg _
+
+lemma abs_nnrpow_two (a : A) : (abs a) ^ (2 : NNReal) = star a * a := by
+  simp only [abs_nonneg, nnrpow_two]
+  apply abs_mul_self
+
+lemma abs_nnrpow_two_mul (a : A) (x : ℝ≥0) :
+    (abs a) ^ (2 * x) = (star a * a) ^ x := by rw [← nnrpow_nnrpow, abs_nnrpow_two]
+
+/-- This and the previous need new names. -/
+lemma abs_nnrpow (a : A) (x : ℝ≥0) :
+    (abs a) ^ x = (star a * a) ^ (x / 2) := by
+  simp only [← abs_nnrpow_two_mul, mul_div_left_comm, ne_eq, OfNat.ofNat_ne_zero,
+    not_false_eq_true, div_self, mul_one]
+
+end abs
+
+variable {A : Type*} [NonUnitalNormedRing A] [StarRing A]
+section IsSelfAdjoint
+
+variable [Module ℝ A] [SMulCommClass ℝ A A] [IsScalarTower ℝ A A]
+variable [NonUnitalContinuousFunctionalCalculus ℝ (IsSelfAdjoint : A → Prop)]
+
+lemma mul_self_eq_mul_self {a : A} (ha : IsSelfAdjoint a := by cfc_tac) : a * a =
+    cfcₙ (fun (x : ℝ) ↦ x * x) a := by
+  conv_lhs => rw [← cfcₙ_id' ℝ a, ← cfcₙ_mul ..]
+section Unique
+
+variable [UniqueNonUnitalContinuousFunctionalCalculus ℝ A]
+
+lemma cfcₙ_sqrt_mul_self {a : A} (ha : IsSelfAdjoint a) :
+    cfcₙ Real.sqrt (a * a) = cfcₙ (fun x ↦ √(x * x)) a := by
+  rw [mul_self_eq_mul_self ha, ← cfcₙ_comp a (f := fun x ↦ x * x) (g := fun x ↦ √x),
+     Function.comp_def]
+
+section NonnegSpectrumClass
+
+variable [PartialOrder A] [StarOrderedRing A] [NonnegSpectrumClass ℝ A]
+
+lemma sqrt_eq_real_sqrt {a : A} (ha : 0 ≤ a := by cfc_tac) :
+    CFC.sqrt a = cfcₙ Real.sqrt a := by
+  rw [sqrt_eq_iff _ (hb := cfcₙ_nonneg (A := A) (fun x _ ↦ Real.sqrt_nonneg x)),
+    ← cfcₙ_mul ..]
+  conv_rhs => rw [← cfcₙ_id (R := ℝ) a]
+  refine cfcₙ_congr fun x hx ↦ ?_
+  refine Real.mul_self_sqrt ?_
+  exact quasispectrum_nonneg_of_nonneg a ha x hx
+
+lemma abs_of_nonneg {a : A} (ha : 0 ≤ a) : abs a = a := by
+  rw [abs, ha.star_eq, sqrt_mul_self a ha]
+
+lemma abs_eq_norm {a : A} (ha : IsSelfAdjoint a) :
+    abs a = cfcₙ (‖·‖) a := by
+   simp only [abs, Real.norm_eq_abs, ← Real.sqrt_sq_eq_abs, sq]
+   rw [sqrt_eq_real_sqrt (star_mul_self_nonneg a), ha.star_eq, cfcₙ_sqrt_mul_self ha]
+
+section CStarRing
+
+variable [CStarRing A]
+
+lemma abs_eq_zero_iff {a : A} : abs a = 0 ↔ a = 0 := by
+  rw [abs, sqrt_eq_zero_iff _, CStarRing.star_mul_self_eq_zero_iff]
```

Comment:
This should really just assume `NonUnitalCStarAlgebra` instead of everything else. And it shouldn't go in this file.

Reply:
This is a good example of where I am puzzled. How are you determining that this is the right assumption? Indeed, there is a giant laundry list of typeclass assumptions I've included to get this result, and the reason for doing that was that if one had a `NonUnitalCStarAlgebra` instance, this instance would "contain" instances of these other things. Is it the case that we want to *exclude* this result from situations where you don't have all of `NonUnitalCStarAlgebra`? I.e. is it explicitly bad to have this result work in situations that contain fewer assumptions? (I also will have to look at this more closely, since you might also be saying that the frankenstein list was actually NOT going to work for `NonUnitalCStarAlgebra` instances because of the weird `NonUnitalNormedRing` assumptions or something like this. Anyhow, I'd like to know what the thinking is, here.)